File: python/src/wh40kdc/imports/newrecruit_simple.py
"""NewRecruit "simple" markdown-ish text adapter.

Shape::

    <breadcrumb> - <faction> - <list name> - [N pts]

    # ++ Army Roster ++ [N pts]
    ## Configuration
    Battle Size: <Label>
    Detachment: <Name>

    ## <Section> [N pts]
    <Unit> [N pts]: <wargear>
    <Unit> [N pts]:
    • <count>x <ModelType>[ [N pts]]: <wargear>

Enhancements are inlined in the wargear list as ``<Name> [N pts]`` — the only
wargear token wearing a ``[…]`` pts suffix. ``Warlord`` and the detachment
"<X> Character" keyword are also stripped from the list and set as flags.
Per-model-type breakdowns under ``•`` lines are collapsed onto the parent
unit.

Python mirror of ``tools/src/import/newrecruit-simple.ts``.
"""

from __future__ import annotations

import re
from typing import Any

from wh40kdc.imports.adapter import FormatAdapter
from wh40kdc.imports.newrecruit_text import (
    classify_wargear_list,
    split_wargear_list,
    strip_parenthetical,
)

# Point brackets may carry comma-separated faction resources after the pts
# figure (e.g. `[4485pts, 29Cabal Points]`); the tail is recognized and
# discarded — only the pts figure is consumed.
_FIRST_LINE = re.compile(r"^(.+)\s-\s\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\]\s*$", re.IGNORECASE)
_ROSTER_HEADER = re.compile(
    r"^#\s*\+\+\s*Army Roster\s*\+\+\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\]\s*$",
    re.IGNORECASE,
)
_ROSTER_HEADER_ANYWHERE = re.compile(r"^#\s*\+\+\s*Army Roster\s*\+\+", re.MULTILINE)
# Some exports omit the `# ++ Army Roster ++` line and open straight with a
# `## Section` heading — accept either marker.
_SECTION_HEADER_ANYWHERE = re.compile(r"^##\s+", re.MULTILINE)
_SECTION_HEADER = re.compile(r"^##\s*(.+?)(?:\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\])?\s*$")
_UNIT_LINE = re.compile(
    r"^(.+?)\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\](?:\s*:\s*(.*))?$", re.IGNORECASE
)
_BULLET = re.compile(
    r"^\s*•\s*(\d+)x\s+(.+?)(?:\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\])?(?:\s*:\s*(.*))?\s*$"
)
_SPLIT_LINES = re.compile(r"\r?\n")

_UNIT_TOTAL_PREFIX = re.compile(r"^Unit total:\s*", re.IGNORECASE)
_ATTACHMENT_TOKEN = re.compile(
    r"^Attachment:\s*(leader|support)\s*->\s*(.+?)(\s+\[provisional\])?$",
    re.IGNORECASE,
)


def _new_unit(name: str, displayed_pts: int | None) -> dict[str, Any]:
    return {
        "raw_name": name,
        "is_character": False,
        "is_warlord": False,
        "keyword_overrides": [],
        "seen_keyword_overrides": set(),
        "enhancement_raw_name": None,
        "enhancement_pts": None,
        "displayed_pts": displayed_pts,
        "saw_bullet": False,
        "model_count": 1,
        "leader_attachment": None,
        # Aggregated wargear, keyed by name (insertion-ordered). Counts sum
        # across `• Nx ModelType` breakdowns.
        "wargear": {},
        "loadout_groups": [],
    }


def _apply_tokens(
    unit: dict[str, Any], tokens_csv: str, multiplier: int = 1
) -> list[dict[str, Any]]:
    wargear_tokens: list[str] = []
    for token in split_wargear_list(tokens_csv):
        attachment = _ATTACHMENT_TOKEN.match(token)
        if attachment:
            unit["leader_attachment"] = {
                "role": attachment.group(1).lower(),
                "bodyguard_raw_name": attachment.group(2),
                "provisional": attachment.group(3) is not None,
            }
        else:
            wargear_tokens.append(token)

    cls = classify_wargear_list(wargear_tokens)
    if cls["is_warlord"]:
        unit["is_warlord"] = True
    if cls["is_character"]:
        unit["is_character"] = True
    for keyword in cls["keyword_overrides"]:
        if keyword not in unit["seen_keyword_overrides"]:
            unit["keyword_overrides"].append(keyword)
            unit["seen_keyword_overrides"].add(keyword)
    if cls["enhancement_raw_name"] and unit["enhancement_raw_name"] is None:
        unit["enhancement_raw_name"] = cls["enhancement_raw_name"]
        unit["enhancement_pts"] = cls["enhancement_points"]
    for w in cls["wargear"]:
        name = w["raw_name"]
        unit["wargear"][name] = unit["wargear"].get(name, 0) + w["count"] * multiplier
    return cls["wargear"]


def _finish_unit(unit: dict[str, Any]) -> dict[str, Any]:
    displayed = unit["displayed_pts"]
    points = (
        None
        if displayed is None
        else displayed - (unit["enhancement_pts"] if unit["enhancement_pts"] is not None else 0)
    )
    return {
        "raw_name": unit["raw_name"],
        "is_character": unit["is_character"],
        **({"keyword_overrides": unit["keyword_overrides"]} if unit["keyword_overrides"] else {}),
        "model_count": unit["model_count"],
        "points": points,
        "is_warlord": unit["is_warlord"],
        "enhancement_raw_name": unit["enhancement_raw_name"],
        "enhancement_points": (
            None if unit["enhancement_raw_name"] is None else unit["enhancement_pts"]
        ),
        "leader_attachment": unit["leader_attachment"],
        "wargear": [{"raw_name": n, "count": c} for n, c in unit["wargear"].items()],
        **({"loadout_groups": unit["loadout_groups"]} if unit["loadout_groups"] else {}),
    }


def _parse_first_line(line: str) -> dict[str, Any] | None:
    m = _FIRST_LINE.match(line)
    if not m:
        return None
    declared_limit = int(m.group(2))
    parts = [s for s in (p.strip() for p in m.group(1).split(" - ")) if s]
    if not parts:
        return None
    return {
        "name": parts[-1],
        "faction": parts[-2] if len(parts) >= 2 else None,
        "declared_limit": declared_limit,
    }


def _matches(decoded: Any) -> bool:
    if not isinstance(decoded, str):
        return False
    first_non_blank = next((line for line in _SPLIT_LINES.split(decoded) if line.strip()), None)
    if not first_non_blank:
        return False
    if not _FIRST_LINE.match(first_non_blank):
        return False
    # Some exports omit the `# ++ Army Roster ++` line and open straight with
    # a `## Section` heading — accept either marker.
    return (
        _ROSTER_HEADER_ANYWHERE.search(decoded) is not None
        or _SECTION_HEADER_ANYWHERE.search(decoded) is not None
    )


def _parse(decoded: Any) -> dict[str, Any]:
    if not isinstance(decoded, str):
        raise ValueError("newrecruit-simple: input is not a string")
    lines = _SPLIT_LINES.split(decoded)

    name = "Imported roster"
    faction_raw_name: str | None = None
    declared_limit: int | None = None
    total_reported: int | None = None
    detachment_raw_names: list[str] = []
    battle_size_raw: str | None = None
    force_disposition_raw_name: str | None = None
    units: list[dict[str, Any]] = []
    current: dict[str, Any] | None = None
    multi_force = False
    section = "preamble"
    enhancement_pts: list[int | None] = []

    def finalize() -> None:
        nonlocal current
        if current is not None:
            enhancement_pts.append(current["enhancement_pts"])
            units.append(_finish_unit(current))
            current = None

    for raw in lines:
        line = raw.strip()
        if not line:
            continue

        # First non-blank line carries `<breadcrumb> - <faction> - <list> - [N pts]`.
        if section == "preamble" and name == "Imported roster":
            first = _parse_first_line(line)
            if first:
                name = first["name"]
                faction_raw_name = first["faction"]
                declared_limit = first["declared_limit"]
                continue

        roster_match = _ROSTER_HEADER.match(line)
        if roster_match:
            total_reported = int(roster_match.group(1))
            continue

        section_match = _SECTION_HEADER.match(line)
        if section_match:
            finalize()
            heading = section_match.group(1).strip().lower()
            if heading == "configuration":
                section = "configuration"
            else:
                section = "units"
                if "allied" in heading:
                    multi_force = True
            continue

        if section == "configuration":
            # Some exports list units directly after Configuration with no units
            # section heading; a `Name [N pts]` line ends the configuration block.
            if _UNIT_LINE.match(line):
                section = "units"
            else:
                idx = line.find(":")
                if idx > 0:
                    key = line[:idx].strip().lower()
                    value = line[idx + 1 :].strip()
                    if key == "battle size":
                        battle_size_raw = value
                    elif key == "list name":
                        name = value
                    elif key == "faction":
                        faction_raw_name = value
                    elif key == "force disposition":
                        force_disposition_raw_name = value
                    elif key == "detachment":
                        # Parenthetical suffixes ("(3 Detachment Points)") are
                        # presentation, not part of the detachment name.
                        detachment_raw_names.append(strip_parenthetical(value))
                continue

        # Unit section. A bullet line extends the *current* unit.
        bullet_match = _BULLET.match(raw)
        if bullet_match and current is not None:
            count = int(bullet_match.group(1))
            # The unit header has no model-count information. Replace its
            # implicit single-model default once, then aggregate every later
            # breakdown, irrespective of whether that breakdown has wargear.
            if not current["saw_bullet"]:
                current["model_count"] = count
                current["saw_bullet"] = True
            else:
                current["model_count"] += count
            # An explicitly empty `:` suffix is a meaningful exact group.
            # Check syntactic presence rather than truthiness so empty groups
            # survive export → import round-trips.
            if bullet_match.group(4) is not None:
                tokens = bullet_match.group(4)
                unit_total = _UNIT_TOTAL_PREFIX.match(tokens) is not None
                group_wargear = _apply_tokens(
                    current,
                    _UNIT_TOTAL_PREFIX.sub("", tokens),
                    1 if unit_total else count,
                )
                if not unit_total:
                    current["loadout_groups"].append(
                        {
                            "model_name": bullet_match.group(2).strip(),
                            "count": count,
                            "wargear": group_wargear,
                        }
                    )
            continue

        unit_match = _UNIT_LINE.match(line)
        if unit_match:
            finalize()
            unit_name = unit_match.group(1).strip()
            pts = int(unit_match.group(2))
            current = _new_unit(unit_name, pts)
            inline_wargear = (unit_match.group(3) or "").strip()
            if inline_wargear:
                _apply_tokens(current, inline_wargear, 1)
            # Leave model_count at the default 1. If `•` bullet lines follow,
            # the bullet handler resets model_count to the summed counts.
            continue

    finalize()

    total_computed = 0
    for i, u in enumerate(units):
        total_computed += u["points"] or 0
        if i < len(enhancement_pts):
            total_computed += enhancement_pts[i] or 0

    return {
        "name": name,
        "generated_by": None,
        "faction_raw_name": faction_raw_name,
        "detachment_raw_names": detachment_raw_names,
        "force_disposition_raw_name": force_disposition_raw_name,
        "battle_size_raw": battle_size_raw,
        "declared_limit": declared_limit,
        "total_reported": total_reported,
        "total_computed": total_computed,
        "units": units,
        "multi_force": multi_force,
    }


newrecruit_simple_adapter = FormatAdapter(id="newrecruit-simple", matches=_matches, parse=_parse)

File: tools/src/import/newrecruit-simple.ts
/**
 * NewRecruit "simple" markdown-ish text adapter.
 *
 * Shape:
 * ```
 * <breadcrumb> - <faction> - <list name> - [N pts]
 *
 * # ++ Army Roster ++ [N pts]
 * ## Configuration
 * Battle Size: <Label>
 * Detachment: <Name>
 * Show/Hide Options: ...
 *
 * ## <Section> [N pts]
 * <Unit> [N pts]: <wargear>
 * <Unit> [N pts]:
 * • <count>x <ModelType>[ [N pts]]: <wargear>
 * ```
 *
 * Enhancements are inlined in the wargear list as `<Name> [N pts]` — the only
 * wargear token wearing a `[…]` pts suffix. `Warlord` and the detachment
 * "<X> Character" keyword are also stripped from the list and set as flags.
 * Per-model-type breakdowns under `•` lines are collapsed onto the parent unit.
 *
 * @packageDocumentation
 */
import type { FormatAdapter } from "./adapter.js";
import type {
  ParsedLeaderAttachment,
  ParsedRoster,
  ParsedUnit,
  ParsedWargear,
} from "./types.js";
import { classifyWargearList, splitWargearList, stripParenthetical } from "./newrecruit-text.js";

// Point brackets may carry comma-separated faction resources after the pts
// figure (e.g. `[4485pts, 29Cabal Points]`); the tail is recognized and
// discarded — only the pts figure is consumed.
const FIRST_LINE = /^(.+)\s-\s\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\]\s*$/i;
const ROSTER_HEADER =
  /^#\s*\+\+\s*Army Roster\s*\+\+\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\]\s*$/i;
const SECTION_HEADER = /^##\s*(.+?)(?:\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\])?\s*$/;
const UNIT_LINE = /^(.+?)\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\](?:\s*:\s*(.*))?$/i;
const BULLET =
  /^\s*•\s*(\d+)x\s+(.+?)(?:\s*\[\s*(\d+)\s*pts?\s*(?:,[^\]]*)?\])?(?:\s*:\s*(.*))?\s*$/u;
const UNIT_TOTAL_PREFIX = /^Unit total:\s*/i;
const ATTACHMENT_TOKEN =
  /^Attachment:\s*(leader|support)\s*->\s*(.+?)(\s+\[provisional\])?$/i;

interface UnitBuilder {
  raw_name: string;
  is_character: boolean;
  is_warlord: boolean;
  keyword_overrides: Set<string>;
  enhancement_raw_name: string | null;
  enhancement_pts: number | null;
  displayed_pts: number | null;
  /** Whether any `• Nx ModelType` breakdown has supplied the unit count. */
  saw_bullet: boolean;
  model_count: number;
  leader_attachment: ParsedLeaderAttachment | null;
  loadout_groups: {
    model_name: string | null;
    count: number;
    wargear: ParsedWargear[];
  }[];
  /** Aggregated wargear, keyed by name. Counts sum across `• Nx ModelType` breakdowns. */
  wargear: Map<string, number>;
}

function newUnit(name: string, displayed_pts: number | null): UnitBuilder {
  return {
    raw_name: name,
    is_character: false,
    is_warlord: false,
    keyword_overrides: new Set(),
    enhancement_raw_name: null,
    enhancement_pts: null,
    displayed_pts,
    saw_bullet: false,
    model_count: 1,
    leader_attachment: null,
    loadout_groups: [],
    wargear: new Map(),
  };
}

function addWargear(unit: UnitBuilder, items: ParsedWargear[]): void {
  for (const { raw_name, count } of items) {
    unit.wargear.set(raw_name, (unit.wargear.get(raw_name) ?? 0) + count);
  }
}

function applyTokens(unit: UnitBuilder, tokensCsv: string, multiplier = 1): ParsedWargear[] {
  const wargearTokens: string[] = [];
  for (const token of splitWargearList(tokensCsv)) {
    const attachment = ATTACHMENT_TOKEN.exec(token);
    if (attachment) {
      unit.leader_attachment = {
        role: attachment[1].toLowerCase() as "leader" | "support",
        bodyguard_raw_name: attachment[2],
        provisional: attachment[3] !== undefined,
      };
    } else {
      wargearTokens.push(token);
    }
  }
  const cls = classifyWargearList(wargearTokens);
  if (cls.is_warlord) unit.is_warlord = true;
  if (cls.is_character) unit.is_character = true;
  for (const keyword of cls.keyword_overrides) unit.keyword_overrides.add(keyword);
  if (cls.enhancement_raw_name && unit.enhancement_raw_name === null) {
    unit.enhancement_raw_name = cls.enhancement_raw_name;
    unit.enhancement_pts = cls.enhancement_points;
  }
  addWargear(
    unit,
    cls.wargear.map((w) => ({
      raw_name: w.raw_name,
      count: w.count * multiplier,
    })),
  );
  return cls.wargear;
}

function finishUnit(unit: UnitBuilder): ParsedUnit {
  const points =
    unit.displayed_pts === null
      ? null
      : unit.displayed_pts - (unit.enhancement_pts ?? 0);
  return {
    raw_name: unit.raw_name,
    is_character: unit.is_character,
    ...(unit.keyword_overrides.size
      ? { keyword_overrides: [...unit.keyword_overrides] }
      : {}),
    model_count: unit.model_count,
    points,
    is_warlord: unit.is_warlord,
    enhancement_raw_name: unit.enhancement_raw_name,
    enhancement_points:
      unit.enhancement_raw_name === null ? null : unit.enhancement_pts,
    leader_attachment: unit.leader_attachment,
    wargear: [...unit.wargear].map(([raw_name, count]) => ({ raw_name, count })),
    ...(unit.loadout_groups.length > 0
      ? { loadout_groups: unit.loadout_groups }
      : {}),
  };
}

function parseFirstLine(line: string): { name: string; faction: string | null; declared_limit: number | null } | null {
  const m = FIRST_LINE.exec(line);
  if (!m) return null;
  const declared_limit = Number.parseInt(m[2], 10);
  const parts = m[1].split(" - ").map((s) => s.trim()).filter((s) => s.length > 0);
  if (parts.length === 0) return null;
  const list_name = parts[parts.length - 1];
  const faction = parts.length >= 2 ? parts[parts.length - 2] : null;
  return { name: list_name, faction, declared_limit };
}

type Section = "preamble" | "configuration" | "units";

export const newRecruitSimpleAdapter: FormatAdapter = {
  id: "newrecruit-simple",

  matches(decoded: unknown): boolean {
    if (typeof decoded !== "string") return false;
    const lines = decoded.split(/\r?\n/);
    const firstNonBlank = lines.find((l) => l.trim().length > 0);
    if (!firstNonBlank) return false;
    if (!FIRST_LINE.test(firstNonBlank)) return false;
    // Some exports omit the `# ++ Army Roster ++` line and open straight with
    // a `## Section` heading — accept either marker.
    return (
      /^#\s*\+\+\s*Army Roster\s*\+\+/m.test(decoded) ||
      /^##\s+/m.test(decoded)
    );
  },

  parse(decoded: unknown): ParsedRoster {
    if (typeof decoded !== "string") {
      throw new Error("newrecruit-simple: input is not a string");
    }
    const lines = decoded.split(/\r?\n/);

    let name = "Imported roster";
    let faction_raw_name: string | null = null;
    let declared_limit: number | null = null;
    let total_reported: number | null = null;
    const detachment_raw_names: string[] = [];
    let battle_size_raw: string | null = null;
    let force_disposition_raw_name: string | null = null;
    const units: ParsedUnit[] = [];
    let current: UnitBuilder | null = null;
    let multi_force = false;
    let section: Section = "preamble";
    const enhancementPts: (number | null)[] = [];

    const finalize = (): void => {
      if (current) {
        enhancementPts.push(current.enhancement_pts);
        units.push(finishUnit(current));
        current = null;
      }
    };

    for (let i = 0; i < lines.length; i += 1) {
      const raw = lines[i];
      const line = raw.trim();
      if (!line) continue;

      // First non-blank line carries `<breadcrumb> - <faction> - <list> - [N pts]`.
      if (section === "preamble" && name === "Imported roster") {
        const first = parseFirstLine(line);
        if (first) {
          name = first.name;
          faction_raw_name = first.faction;
          declared_limit = first.declared_limit;
          continue;
        }
      }

      const rosterMatch = ROSTER_HEADER.exec(line);
      if (rosterMatch) {
        total_reported = Number.parseInt(rosterMatch[1], 10);
        continue;
      }

      const sectionMatch = SECTION_HEADER.exec(line);
      if (sectionMatch) {
        finalize();
        const heading = sectionMatch[1].trim().toLowerCase();
        if (heading === "configuration") {
          section = "configuration";
        } else {
          section = "units";
          if (heading.includes("allied")) multi_force = true;
        }
        continue;
      }

      if (section === "configuration") {
        // Some exports list units directly after Configuration with no units
        // section heading; a `Name [N pts]` line ends the configuration block.
        if (UNIT_LINE.test(line)) {
          section = "units";
        } else {
          const idx = line.indexOf(":");
          if (idx > 0) {
            const key = line.slice(0, idx).trim().toLowerCase();
            const value = line.slice(idx + 1).trim();
            if (key === "battle size") battle_size_raw = value;
            else if (key === "list name") name = value;
            else if (key === "faction") faction_raw_name = value;
            else if (key === "force disposition") {
              force_disposition_raw_name = value;
            } else if (key === "detachment") {
              // Parenthetical suffixes ("(3 Detachment Points)") are
              // presentation, not part of the detachment name.
              detachment_raw_names.push(stripParenthetical(value));
            }
          }
          continue;
        }
      }

      // Unit section. A bullet line extends the *current* unit.
      const bulletMatch = BULLET.exec(raw);
      if (bulletMatch && current) {
        const count = Number.parseInt(bulletMatch[1], 10);
        // The unit header has no model-count information. Replace its implicit
        // single-model default once, then aggregate every later breakdown,
        // irrespective of whether that breakdown has wargear.
        if (!current.saw_bullet) {
          current.model_count = count;
          current.saw_bullet = true;
        } else {
          current.model_count += count;
        }
        // An explicitly empty `:` suffix is a meaningful exact group. Check
        // for syntactic presence rather than its truthiness so empty groups
        // survive export → import round-trips.
        if (bulletMatch[4] !== undefined) {
          const unitTotal = UNIT_TOTAL_PREFIX.test(bulletMatch[4]);
          const groupWargear = applyTokens(
            current,
            bulletMatch[4].replace(UNIT_TOTAL_PREFIX, ""),
            unitTotal ? 1 : count,
          );
          if (!unitTotal) {
            current.loadout_groups.push({
              model_name: bulletMatch[2].trim(),
              count,
              wargear: groupWargear,
            });
          }
        }
        continue;
      }

      const unitMatch = UNIT_LINE.exec(line);
      if (unitMatch) {
        finalize();
        const unitName = unitMatch[1].trim();
        const pts = Number.parseInt(unitMatch[2], 10);
        current = newUnit(unitName, pts);
        const inlineWargear = unitMatch[3]?.trim() ?? "";
        if (inlineWargear.length > 0) {
          applyTokens(current, inlineWargear, 1);
        }
        // Leave model_count at the default 1. If `•` bullet lines follow, the
        // bullet handler resets model_count to the (summed) bullet counts.
        continue;
      }
    }
    finalize();

    let total_computed = 0;
    for (let i = 0; i < units.length; i += 1) {
      total_computed += units[i].points ?? 0;
      total_computed += enhancementPts[i] ?? 0;
    }

    return {
      name,
      generated_by: null,
      faction_raw_name,
      detachment_raw_names,
      battle_size_raw,
      force_disposition_raw_name,
      declared_limit,
      total_reported,
      total_computed,
      units,
      multi_force,
    };
  },
};
